### PR TITLE
Relist broker catalogs based on when broker became ready and relist interval

### LINKED ISSUE #665: Integration tests flaking, timing out waiting for broker to come up
Opening a separate issue for this flake, as #643 started to become a general hub for all flaky behavior.

Integration tests are intermittently failing with error:

```
[Build & Unit/Integration Tests] --- FAIL: TestBasicFlows (40.09s)
[Build & Unit/Integration Tests]     controller_test.go:113: error waiting for broker to become ready: timed out waiting for the condition
```

### PR DESCRIPTION
**Pull request by pmorie**
Part of #627; depends on #630.  Fixes #665.

This PR:

1.  Adds a new broker relist interval flag to the controller
2.  Only relists a ready broker's catalog when it has been longer than the relist interval since the broker became ready

I **really** want this to be in 0.0.3

### CODE REVIEW

**Comment by pmorie** [comment]
cc @derekwaynecarr

File: pkg/controller/controller.go [hunk 1]
```diff
@@ -155,15 +159,45 @@ const (
...
+// - the broker's deletion timestamp is set
+// - the broker has no status conditions
+// - the broker has a ready condition whose status is true, and it has been
+//   longer than the given relist interval since the broker became ready
+func shouldReconcileBroker(broker *v1alpha1.Broker, now time.Time, relistInterval time.Duration) bool {
+	if broker.DeletionTimestamp != nil || len(broker.Status.Conditions) == 0 {
+		return true
+	}
+
+	for _, condition := range broker.Status.Conditions {
```
Comment (MHBauer) [inline pkg/controller/controller.go]: Why loop through the conditions? Shouldn't we relist if it's not Ready as well?

The only case we shouldn't relist is ready =true and time is not enough. Everything else we need to go do something.

File: pkg/controller/controller.go [hunk 2]
```diff
@@ -155,15 +159,45 @@ const (
...
+
+	return false
+}
+
 // reconcileBroker is the control-loop that reconciles a Broker.
 func (c *controller) reconcileBroker(broker *v1alpha1.Broker) {
 	glog.V(4).Infof("Processing Broker %v", broker.Name)
 
+	if !shouldReconcileBroker(broker, time.Now(), c.brokerRelistInterval) {
+		glog.V(10).Infof("Not processing Broker %v because the relist interval has not elapsed since the broker became ready", broker.Name)
```
Comment (MHBauer) [inline pkg/controller/controller.go]: Based on the log message is this the only case will return false?

File: pkg/controller/controller.go [hunk 3]
```diff
@@ -155,15 +159,45 @@ const (
...
+		if condition.Type == v1alpha1.BrokerConditionReady {
+			if condition.Status == v1alpha1.ConditionTrue {
+				return now.After(condition.LastTransitionTime.Add(relistInterval))
+			}
+
+			return true
+		}
+	}
+
+	return false
```
Comment (MHBauer) [inline pkg/controller/controller.go]: What is the case that results in this false?

File: charts/catalog/values.yaml [hunk 4]
```diff
@@ -54,3 +54,5 @@ controllerManager:
   verbosity: 10
   # Resync interval; format is a duration (`20m`, `1h`, etc)
   resyncInterval: 5m
+  # Broker relist interval; format is a duration (`20m`, `1h`, etc)
+  brokerRelistInterval: 30m
```
Comment (MHBauer) [inline charts/catalog/values.yaml]: Why set a default of 24 hours in the code and then 30minutes here?

File: cmd/controller-manager/app/options/options.go [hunk 5]
```diff
@@ -36,6 +36,7 @@ type ControllerManagerServer struct {
 }
 
 const defaultResyncInterval = 5 * time.Minute
+const defaultBrokerRelistInterval = 24 * time.Hour
```
Comment (simonleung8) [inline cmd/controller-manager/app/options/options.go]: Default is 24hrs, but doc says 30 mins. Potentially surprising to people if they don't have a default set in the chart?

**Comment by pmorie** [comment]
@arschles @vaikas-google @duglin @MHBauer @krancour I know a lot of poeple have friday off, but if possible I would like this to go into the 0.0.3 release or at least be merged before I go under the knife 🔪

File: cmd/controller-manager/app/options/options.go (hunk 5 above)
Comment (pmorie) [inline cmd/controller-manager/app/options/options.go]: It's set to 30 minutes by the helm chart (see `values.yaml`), which is a more reasonable default for development.  When you run the binary, I think the default should be more reasonable for production, hence 24 hours.

File: pkg/controller/controller.go (hunk 3 above)
Comment (pmorie) [inline pkg/controller/controller.go]: I'll add comments :)

File: pkg/controller/controller.go (hunk 2 above)
Comment (pmorie) [inline pkg/controller/controller.go]: No, I'll change the comment.

File: charts/catalog/values.yaml (hunk 4 above)
Comment (pmorie) [inline charts/catalog/values.yaml]: 24 hours is a better default for production, 30 minutes is better for dev.  It seems to be throwing people, so I'll just make them the same.

**Comment by pmorie** [comment]
For the record, I don't intend this to be my statement on the final word of correct behavior.  I think the eventual things users will expect are:

1.  For a particular broker, I can determine what relist behavior I want from a broker.  Sometimes I'll want periodic relist, sometime's i'll want only manual relist
2.  There should be a porcelain command to manually relist the broker that works whatever the relist policy is

My goal with this PR was just:

1.  Make it less spammy than it is now
2.  Fix flake #665

File: pkg/controller/controller.go (hunk 2 above)
Comment (pmorie) [inline pkg/controller/controller.go]: Actually yes, you found a bug.

**Comment by pmorie** [comment]
@MHBauer 

> Do we have a test to make sure this works? Set the timeout really long, do a reconcile/whatever, make sure nothing was reconciled.

I have unit tests that exercise the logic for skipping.  I think an e2e or integration test for this would be good, but I want to punt for now because of the general stability issues with integration that I would like to work through before adding a new, potentially racy test, and we don't have e2es yet.

File: cmd/controller-manager/app/options/options.go (hunk 5 above)
Comment (pmorie) [inline cmd/controller-manager/app/options/options.go]: I changed it to match the default set in the code :)

File: cmd/controller-manager/app/options/options.go [hunk 6]
```diff
@@ -69,5 +71,4 @@ func (s *ControllerManagerServer) AddFlags(fs *pflag.FlagSet) {
 	fs.StringVar(&s.ServiceCatalogAPIServerURL, "service-catalog-api-server-url", "", "The URL for the service-catalog API server")
 	fs.StringVar(&s.ServiceCatalogKubeconfigPath, "service-catalog-kubeconfig", s.ServiceCatalogKubeconfigPath, "Path to service-catalog kubeconfig")
 	fs.DurationVar(&s.ResyncInterval, "resync-interval", s.ResyncInterval, "The interval on which the controller will resync its informers")
-	fs.BoolVar(&s.OSBAPIContextProfile, "enable-osb-api-context-profile", s.OSBAPIContextProfile, "Whether or not to send the proposed optional OpenServiceBroker API Context Profile field")
```
Comment (MHBauer) [inline cmd/controller-manager/app/options/options.go]: weird rebase?

Comment (MHBauer) [inline cmd/controller-manager/app/options/options.go]: I think you were trying to add your flag here.

Comment (MHBauer) [inline cmd/controller-manager/app/options/options.go]: your => "the new broker relist interval"

**Comment by MHBauer** [comment]
sorry, I literally fell on the keyboard.

**Pull request merged**